lib/accesslists/hierarchy.go

```diff
@@ -440,23 +441,204 @@ func UserMeetsRequirements(identity types.User, requires accesslist.Requires) bo
+		// e.g. A has nested ownership B, B has nested membership C, C has direct members like alice
+		ownerHierarchy = append(ownerHierarchy, accessList)
+	}
+	ok, err := s.validMembership(ctx, accessList, user)
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+	if !ok {
+		// User is not a valid member of the starting access list
+		// return current owner hierarchy (not empty if a user is direct ownership)
+		// and empty member hierarchy
+		return memberHierarchy, ownerHierarchy, nil
+	}
+	// User is a valid member of the starting access list
+	// Include access list in the member hierarchy
+	memberHierarchy = append(memberHierarchy, accessList)
+
+	// Fetch ancestors via MemberOf edges while checking user requirements
+	ancestors, err := s.membershipAncestors(ctx, accessList, user)
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+	// Expand ancestors via OwnerOf edges while checking user requirements
+	owners, err := s.expandOwnerOf(ctx, ancestors, user)
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+	return append(memberHierarchy, ancestors...), append(ownerHierarchy, owners...), nil
+}
+
+func (s *Hierarchy) validMembership(ctx context.Context, list *accesslist.AccessList, user types.User) (bool, error) {
+	m, err := s.AccessListsService.GetAccessListMember(ctx, list.GetName(), user.GetName())
+	if err != nil {
+		if trace.IsNotFound(err) {
+			return false, nil
+		}
+		return false, trace.Wrap(err)
+	}
+	if m.IsExpired(s.Clock) || !UserMeetsRequirements(user, list.GetMembershipRequires()) {
+		return false, nil
+	}
+	return true, nil
+}
+
+// membershipAncestors fetches ancestors via MemberOf edges while checking user requirements.
+func (s *Hierarchy) membershipAncestors(ctx context.Context, list *accesslist.AccessList, user types.User) ([]*accesslist.AccessList, error) {
+	return getAncestors(ctx, list, RelationshipKindMember, s.AccessListsService, withUserRequirementsCheck(user, s.Clock))
+}
+
+func (s *Hierarchy) expandOwnerOf(ctx context.Context, ancestors []*accesslist.AccessList, user types.User) ([]*accesslist.AccessList, error) {
+	var out []*accesslist.AccessList
+	for _, v := range ancestors {
+		for _, name := range v.Status.OwnerOf {
+			lst, err := s.AccessListsService.GetAccessList(ctx, name)
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+			if UserMeetsRequirements(user, lst.GetOwnershipRequires()) {
+				out = append(out, lst)
+			}
+		}
+	}
+	return out, nil
+}
+
+func (s *Hierarchy) validDirectOwner(user types.User, acl *accesslist.AccessList) bool {
+	if !UserMeetsRequirements(user, acl.GetOwnershipRequires()) {
+		return false
+	}
+	for _, v := range acl.Spec.Owners {
+		if v.Name == user.GetName() {
+			return true
+		}
+	}
+	return false
+}
+
+func getAncestors(ctx context.Context, accessList *accesslist.AccessList, kind RelationshipKind, g AccessListAndMembersGetter, opts ...ancestorOption) ([]*accesslist.AccessList, trace.Error) {
```

Comment:
So now we have that. Do we need separate:

```
func collectOwners(ctx context.Context, accessList *accesslist.AccessList, g AccessListAndMembersGetter, owners map[string]*accesslist.Owner, visited map[string]struct{}) error {
```

?

Reply:
Yea, I think that The `collectAncestors` can be leverage here with small alignment. Will add this as TODO and rise a separator PR to not mix refactoring with the `GetHierarchyForUser` core PR implementation.